python/ray/experimental/workflow/storage.py

```diff
@@ -0,0 +1,94 @@
+import json
+import pathlib
+from typing import Any, Union
+import uuid
+
+import ray
+import ray.cloudpickle
+import ray._private.utils
+
+# TODO(suquark): in the future we may use general URLs to support
+# different storages, e.g. "s3://xxxx/xxx". Currently we just use
+# 'pathlib.Path' for convenience.
+
+
+class Storage:
+    """Base class for accessing workflow storage.
+
+    TODO(suquark): Support S3 etc. in the future. Currently only support
+    local/shared filesystems.
+    """
+
+    def __init__(self, workflow_root_dir: Union[pathlib.Path, str]):
+        self._workflow_root_dir = pathlib.Path(workflow_root_dir)
+
+    def _write_atomic(self, writer, mode, obj, path: pathlib.Path, overwrite):
+        # TODO(suquark): race conditions like two processes writing the
+        # same file is still not safe. This may not be an issue, because
+        # in our current implementation, we only need to guarantee the
+        # file is either fully written or not existing.
+        fullpath = self._workflow_root_dir / path
+        if fullpath.exists() and not overwrite:
+            raise FileExistsError(fullpath)
+        tmp_new_filename = fullpath.with_suffix(fullpath.suffix + "." +
+                                                uuid.uuid4().hex)
+        with open(tmp_new_filename, mode) as f:
+            writer(obj, f)
+        if fullpath.exists():
+            backup_path = fullpath.with_suffix(fullpath.suffix + ".backup")
```

Comment:
In which case will we use backup?

Reply:
when overwriting the old files but fail in the middle, so the new file does not exist and we can use backup files